ctest-next/src/template.rs

```diff
@@ -291,4 +458,139 @@ impl<'a> TranslateHelper<'a> {
 
         Ok(self.generator.map(item))
     }
+
+    /// Recursively get the properly mapped type for some syn::Type.
+    ///
+    /// This method is meant to only be used to translate simpler types, such as primitives or
+    /// pointers/references to primitives.
+    fn c_type_raw(&self, ty: &syn::Type) -> Result<String, TranslationError> {
+        let type_name = match ty {
+            syn::Type::Path(p) => p.path.segments.last().unwrap().ident.to_string(),
+            syn::Type::Ptr(p) => self.c_type_raw(&p.elem)?,
+            syn::Type::Reference(r) => self.c_type_raw(&r.elem)?,
+            _ => ty.to_token_stream().to_string(),
+        };
+
+        let unmapped_c_type = self.translator.translate_type(ty)?;
+        let item = if self.ffi_items.contains_struct(&type_name) {
+            MapInput::StructType(&unmapped_c_type)
+        } else if self.ffi_items.contains_union(&type_name) {
+            MapInput::UnionType(&unmapped_c_type)
+        } else {
+            MapInput::Type(&unmapped_c_type)
+        };
+
+        Ok(self.generator.map(item))
+    }
+
+    /// Partially translate a Rust bare function type into its equivalent C type.
+    ///
+    /// It returns the translated return type, translated argument types, and whether
+    /// it is variadic as a tuple.
+    pub(crate) fn translate_signature_partial(
+        &self,
+        signature: &syn::TypeBareFn,
+    ) -> Result<(String, Vec<String>, bool), TranslationError> {
+        let args = signature
+            .inputs
+            .iter()
+            .map(|arg| self.c_type_raw(&arg.ty))
+            .collect::<Result<Vec<_>, TranslationError>>()?;
+        let return_type = match &signature.output {
+            syn::ReturnType::Default => "void".to_string(),
+            syn::ReturnType::Type(_, ty) => match ty.deref() {
+                syn::Type::Never(_) => "void".to_string(),
+                syn::Type::Tuple(tuple) if tuple.elems.is_empty() => "void".to_string(),
+                _ => self.c_type_raw(ty.deref())?,
+            },
+        };
+        Ok((return_type, args, signature.variadic.is_some()))
+    }
+
+    /// Modify function signatures to properly return pointer types in C.
+    ///
+    /// In C, function pointers and arrays have a different syntax to return them,
+    /// and this translation is done by this method.
+    pub(crate) fn translate_signature_returning_ptr(
+        &self,
+        signature: &str,
+        return_type: &syn::Type,
+    ) -> Result<String, TranslationError> {
```

Comment:
This can be a general cdecl generator right, not specific to signatures anymore right? Maybe `translate_signature_returning_ptr` -> `make_cdecl`, `signature` -> `name`, and `return_type` -> `ty`.